src/Compiler/Checking/SignatureConformance.fs

```diff
@@ -352,14 +354,22 @@ type Checker(g, amap, denv, remapInfo: SignatureRepackageInfo, checkingSig) =
             else 
                 let implTypars, implValTy = implVal.GeneralizedType
                 let sigTypars, sigValTy = sigVal.GeneralizedType
-                if implTypars.Length <> sigTypars.Length then (err {denv with showTyparBinding=true} FSComp.SR.ValueNotContainedMutabilityParameterCountsDiffer) else
-                let aenv = aenv.BindEquivTypars implTypars sigTypars 
-                checkTypars m aenv implTypars sigTypars &&
-                if not (typeAEquiv g aenv implValTy sigValTy) then err denv FSComp.SR.ValueNotContainedMutabilityTypesDiffer
-                elif not (checkValInfo aenv (err denv) implVal sigVal) then false
-                elif implVal.IsExtensionMember <> sigVal.IsExtensionMember then err denv FSComp.SR.ValueNotContainedMutabilityExtensionsDiffer
-                elif not (checkMemberDatasConform (err denv) (implVal.Attribs, implVal, implVal.MemberInfo) (sigVal.Attribs, sigVal, sigVal.MemberInfo)) then false
-                else checkAttribs aenv implVal.Attribs sigVal.Attribs (fun attribs -> implVal.SetAttribs attribs)              
+                if implTypars.Length <> sigTypars.Length then (err {denv with showTyparBinding=true} FSComp.SR.ValueNotContainedMutabilityParameterCountsDiffer) 
+                else
+                    let aenv = aenv.BindEquivTypars implTypars sigTypars 
+                    checkTypars m aenv implTypars sigTypars &&
+                        let strictTyEquals = typeAEquiv g aenv implValTy sigValTy
+                        let onlyDiffersInNullness = not(strictTyEquals) && g.checkNullness && typeAEquiv g {aenv with NullnessMustEqual = false} implValTy sigValTy
+
+                        // The types would be equal if we did not have nullness checks => lets just generate a warning, not an error
+                        if onlyDiffersInNullness then
```

Comment:
Shouldn't the signature be at least as restrictive as the implementation.

E.g. If the implementation can produce null then if the signature says, never produces null, wouldn't that be a problem?

Although I could easily misunderstand the feature :-)

Reply:
The nullness difference can be on any level - argument, return type, generic parameter to either of them, domainTy vs retTy to a function argument.
It really then depends on the co- vs contra-variance of that argument which is a concept that I did not want to introduce just for this check.

That it's why it checks for equality.

Reply:
I think that may make sense.